### DISCUSSION TITLE
C# output to console

### DISCUSSION THREAD

**Discussion by Declan1984** [TabularEditor/TabularEditor3#578] [🙋♀️TE3 Q&A] [answered]
When running a C# script, is there a way to print out to console to view output for debugging, etc.?  It appears that the standard method is Console.WriteLine() but this doesn't appear to output anywhere.  I have been using the .Output() function on string variables which displays output but in a message window so it has to be clicked through for each item.

**Comment by otykier** [comment, marked as answer]
There is no console in TE3 as it is not a console application. Instead, consider appending to a string and then output the string at the end of the script, or write your output directly to a file on disk.

**Reply by Declan1984** [reply]
Thanks, if possible, some type of console output in the application or a way to execute the script from another terminal or call into the model would be very helpful.  The concatenated string is a good option but a little more clunky.  The script is a powerful feature of TE3 but debugging needs to be easier to make it better.

**Reply by otykier** [reply]
We do have plans for c# script debugging in TE3 (allowing you to insert breakpoints, inspect variables, etc. similar to Visual Studio), although no timeline yet. If you need console execution of scripts, consider using TE2, [which has a CLI that lets you run scripts](https://docs.tabulareditor.com/te2/Command-line-Options.html). TE2 exposes the same object model for c# scripts as TE3, with just a few minor differences. When a script is run in CLI mode, the [`Info()`](https://docs.tabulareditor.com/api/TabularEditor.Shared.Scripting.ScriptHost.html#TabularEditor_Shared_Scripting_ScriptHost_Info_System_String_), [`Warning()`](https://docs.tabulareditor.com/api/TabularEditor.Shared.Scripting.ScriptHost.html#TabularEditor_Shared_Scripting_ScriptHost_Warning_System_String_) and [`Error()`](https://docs.tabulareditor.com/api/TabularEditor.Shared.Scripting.ScriptHost.html#TabularEditor_Shared_Scripting_ScriptHost_Error_System_String_) methods output to the console instead of to a dialog box, or you can use `Console.WriteLine()` as usual.

**Reply by Declan1984** [reply]
That's great to hear, I would vote for a debugging feature.  One of our team's goals is to fully script out the building of the model so our model can be re-built from code and not just the model json objects so we have some re-usability or ability to deploy to different environments.

**Reply by otykier** [reply]
C# script debugging is part of a more comprehensive feature that allows you to use TE3 as an IDE for C# applications (the use case is of course Tabular models, but in principle, you could use it for anything). In addition to debugging, we will also allow the creation of code modules that can be reused across scripts.

For now, you may find this [blog post interesting](https://www.esbrina-ba.com/bringing-tabular-editor-c-scripts-to-the-next-level/), which describes how you can bring TE scripting over to your favorite IDE.
